pkg/services/dashboards/database/database.go

```diff
@@ -994,16 +997,220 @@ func (d *dashboardStore) FindDashboards(ctx context.Context, query *dashboards.F
+// query.
+// SQLBuilder is expected to be embedded into the data struct passed into the
+// template. In a template one can use {{.Arg .Foo}} and it will replace that
+// with a "?" placeholder appending the value of .Foo to the Params slice.
+type SQLBuilder struct {
+	Params []any
+}
+
+func (sb *SQLBuilder) Arg(v any) (sql string) { sb.Params = append(sb.Params, v); return "?" }
+
+var searchSQL = template.Must(template.New("").Parse(`
+{{- /*
+####
+#### Here come different implementations for the innermost full-text search query.
+#### A query is expected to return a list of entities with their (OrgID, UID, Kind, Title, ParentFolderUID)
+####
+*/ -}}
+{{- define "text-search" -}}
+SELECT org_id, 'dashboard' as kind, uid, id, folder_uid as parent_uid, title FROM dashboard WHERE title {{ .Dialect.LikeStr }} {{ .Arg .Text }} AND NOT is_folder
+UNION ALL
+SELECT org_id, 'folder' as kind, uid, 0, parent_uid, title FROM folder WHERE title {{ .Dialect.LikeStr }} {{ .Arg .Text }}
+{{ end }}
+{{- define "text-search-mysql-fts" -}}
+SELECT org_id, 'dashboard' as kind, uid, id, folder_uid as parent_uid, title FROM dashboard WHERE MATCH(title) AGAINST({{ .Arg .Text }}) AND NOT is_folder
+UNION ALL
+SELECT org_id, 'folder' as kind, uid, 0, parent_uid, title FROM folder WHERE MATCH(title) AGAINST({{ .Arg .Text }})
+{{ end }}
+{{- define "text-search-legacy" -}}
+SELECT org_id, 'dashboard' as kind, uid, id, folder_uid as parent_uid, title FROM dashboard WHERE title {{ .Dialect.LikeStr }} {{ .Arg .Text }} AND NOT is_folder
+UNION ALL
+SELECT org_id, 'folder' as kind, uid, id, folder_uid as parent_uid, title FROM dashboard WHERE title {{ .Dialect.LikeStr }} {{ .Arg .Text }} AND is_folder
+{{ end }}
+
+SELECT DISTINCT
+	entity.org_id, entity.kind, entity.uid, entity.id, entity.title,
+	dashboard_tag.term as term, f1.uid as folder_uid, f1.title as folder_title FROM (
+	{{- template "text-search-legacy" . -}}
+) AS entity
+LEFT JOIN folder f1 ON (entity.parent_uid = f1.uid AND entity.org_id = f1.org_id)
+{{ if ne .UserID 0 }}   {{- /* If a user is not admin: need to filter by permissions */ -}}
+LEFT JOIN folder f2 ON (f1.parent_uid     = f2.uid AND f1.org_id     = f2.org_id)
+LEFT JOIN folder f3 ON (f2.parent_uid     = f3.uid AND f2.org_id     = f3.org_id)
+LEFT JOIN folder f4 ON (f3.parent_uid     = f4.uid AND f3.org_id     = f4.org_id)
+INNER JOIN permission p ON 
+p.role_id IN (
+	SELECT ur.role_id FROM user_role AS ur    WHERE ur.user_id = {{ .Arg .UserID }} AND (ur.org_id = {{ .Arg .OrgID }} OR ur.org_id = 0)
+	UNION
+	SELECT tr.role_id FROM team_role AS tr    WHERE tr.team_id IN (
+		{{- range $i, $el := .TeamIDs -}}
+		{{- if $i -}},{{- end -}}
+		{{- $.Arg $el -}}
+		{{- end -}}
+	) AND tr.org_id = {{ .Arg .OrgID }}
+	UNION
+	SELECT br.role_id FROM builtin_role AS br WHERE br.role IN ({{ .Arg .Role }}) AND (br.org_id = {{ .Arg .OrgID }} OR br.org_id = 0)
+) AND (
+{{ if .HasSplitScopes }}
```

Comment:
I think it should be safe to assume that we can enabled split scopes if we're going ahead with this new query and querying with split scopes is way more effective. The only reason why we left it behind a feature toggle is that we didn't manage to make them work as well as we hoped.

Reply:
Changed the PR to enable alternative search query only if `splitScopes` is enabled, too.